### PR TITLE
deploy.pypi.sh: Enhance travis to upload wheel

### PR DESCRIPTION
**Pull request by Shade5**
Closes https://github.com/coala/coala/issues/3459

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### CODE REVIEW

**Comment by sils** [comment]
WARNING: note that rultor will not fail to merge this even if this change is erroneous! We gotta be very careful. However if it is erroneous it will fail all future merges since rultor always uses the config from the base branch.

**Comment by Shade5** [comment]
Is this change erroneous? Also, I'm assuming the username and password for testpypi are the same as the ones used for pypi

**Comment by sils** [comment]
well that's the thing, rultor is hard to debug and predict, you don't necessarily know if there are problems

**Comment by Shade5** [comment]
So should I modify Travis? From what I understood I keep rultor as it is but modify Travis to upload to testpypi. Am I correct?

**Comment by Shade5** [comment]
Another problem is that the credentials are encrypted by rultor and can only be decrypted by rultor

**Comment by Shade5** [comment]
@sils I've tried to expose the secrets from a fork. It's not possible as mentioned [here](https://docs.travis-ci.com/user/encrypting-files/).
You can go ahead and create a secrets.txt.enc for travis when you're ready.

**Comment by sils** [comment]
@Shade5 travis has the following environment variables exposed now: `PYPI_USER`, `PYPI_PASSWORD`, they allow to access pypi and IIUC they're not present on builds from PRs from forks.

File: .misc/deploy.testpypi.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+# Ship it!
+echo "Uploading coala to test-pypi"
+pip3 install twine wheel
+python3 setup.py sdist bdist_wheel
+# Upload one by one to avoid timeout
+# Don't block merging because pypi is in error 500 mood again
+set +e
+twine upload dist/*.whl -r https://testpypi.python.org/pypi -u "$PYPI_USER" -p "$PYPI_PASSWORD"
```
Comment (sils) [inline .misc/deploy.testpypi.sh]: https://travis-ci.org/coala/coala/jobs/194017686#L1064

**Comment by Shade5** [comment]
Getting this error, I've registered and am trying to upload through my credentials
```
a@a ~/workspace/coala/coala $ twine upload dist/*.whl --config-file ./.pypirc -r test -u "$PYPI_USER" -p "$PYPI_PASSWORD"
Uploading distributions to https://testpypi.python.org/pypi
Uploading coala-0.10.0.dev99999999999999-py34.py35.py36-none-any.whl
[================================] 191268/191268 - 00:00:05
HTTPError: 403 Client Error: You are not allowed to edit 'coala' package information for url: https://testpypi.python.org/pypi
```
So if i'm not allowed, shouldn't Travis be allowed? But its failing too

**Comment by sils** [comment]
you're doing builds from a fork and generally people working from a fork might do things like exposing keys so your build probably doesn't have the variables exposed

**Comment by Shade5** [comment]
So Travis will always fail to upload to testpypi until it is merged? What can I do?

**Comment by sils** [comment]
@Shade5 I don't have time for this discussion, I'm involved with too many things, please talk to @Mixih the release team leader

**Comment by Mixih** [comment]
@Shade5 That is intended behavior. It should fail until merge, otherwise we upload a bunch of broken packages to the pypi.

**Comment by Shade5** [comment]
So can't test without merging?

**Comment by jayvdb** [comment]
@shade5, this is about publishing on the test server.

The pypi distribution name can be customised, like `coala-shade5`, so that anyone can test packaging using their own testpypi account, without any permission from anyone.

File: .misc/deploy.testpypi.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# Get environment variables, readily decrypted by rultor
+
+
+# Ship it!
+echo "Uploading coala to test-pypi"
+pip3 install twine wheel
+python3 setup.py sdist bdist_wheel
+# Upload one by one to avoid timeout
+# Don't block merging because pypi is in error 500 mood again
+set +e
```
Comment (jayvdb) [inline .misc/deploy.testpypi.sh]: this isnt good enough. It should gracefully fail only if it is a server error(500), and not if it is any error.
See https://github.com/coala/coala/issues/3644

File: .misc/deploy.testpypi.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+twine upload dist/*.whl -r test -u "$PYPI_USER" -p "$PYPI_PASSWORD"
+set -e
+
```
Comment (jayvdb) [inline .misc/deploy.testpypi.sh]: Trailing blank line.

File: .travis.yml [hunk 4]
```diff
@@ -62,6 +62,7 @@ script:
     else
       set -e
       bash .misc/tests.sh
+      bash .misc/deploy.testpypi.sh
```
Comment (jayvdb) [inline .travis.yml]: This is a deploy step; it should only happen if all checks pass

File: .pypirc [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[distutils]
+index-servers=
```
Comment (jayvdb) [inline .pypirc]: When release manager tries to do a production release, will this cause a problem?
Will it override their own rc file.

File: .misc/deploy.testpypi.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (jayvdb) [inline .misc/deploy.testpypi.sh]: Imo you should improve deploy.pypi.sh , maybe using two env vars to control

1. Whether test or production is used,
2. Whether rultor secrets are needed

Comment (Shade5) [inline .misc/deploy.testpypi.sh]: Both travis and rultor run deploy.pypi.sh, and actions depends on who's calling them?
Wouldn't that cause problems later?

File: .pypirc (hunk 5 above)
Comment (Shade5) [inline .pypirc]: No, the default is pypi. 
It is sent to testpypi only if it is specified.

File: .misc/deploy.testpypi.sh (hunk 2 above)
Comment (Shade5) [inline .misc/deploy.testpypi.sh]: Should I fix it here or leave it to #3644 ?

File: .misc/deploy.testpypi.sh (hunk 6 above)
Comment (jayvdb) [inline .misc/deploy.testpypi.sh]: What problems?

File: .misc/deploy.testpypi.sh (hunk 2 above)
Comment (jayvdb) [inline .misc/deploy.testpypi.sh]: this is hiding bugs in your PR. It needs to be fixed.

File: .travis.yml [hunk 7]
```diff
@@ -62,6 +62,9 @@ script:
     else
       set -e
       bash .misc/tests.sh
+      if [ $? -eq 0 ]; then
```
Comment (jayvdb) [inline .travis.yml]: No. All checks. The 'script' section must finish.
See https://docs.travis-ci.com/user/deployment

File: .misc/deploy.testpypi.sh (hunk 6 above)
Comment (Shade5) [inline .misc/deploy.testpypi.sh]: I've added what you suggested in the latest commit, it's just a test.
But isn't depending on environment variables a bit shakey?

**Comment by Shade5** [comment]
I'm sorry, but I can't figure out how to fix the `appveyor` problem.
Stuck at installing `libxml2` and `libxslt`

File: .misc/deploy.pypi.sh [hunk 8]
```diff
@@ -2,22 +2,40 @@
...
-set +e
-twine upload dist/*.whl -u "$PYPIUSER" -p "$PYPIPW"
-twine upload dist/*.tar.gz -u "$PYPIUSER" -p "$PYPIPW"
-set -e
+if ! [ -z "$PYPIUSER" ]
+then
+    echo "In RULTOR"
+    # Ship it!
+    echo "Uploading coala to pypi"
+    pip3 install twine wheel
```
Comment (jayvdb) [inline .misc/deploy.pypi.sh]: Please remove the duplication.

File: .misc/deploy.pypi.sh [hunk 9]
```diff
@@ -2,22 +2,40 @@
 
 # Do not use `set -x` here as then it displays the PYPIPW in logs
 set -e
-
```
Comment (jayvdb) [inline .misc/deploy.pypi.sh]: Why remove this line? Please read you diff before uploading for review.

File: .misc/deploy.pypi.sh [hunk 10]
```diff
@@ -2,22 +2,40 @@
...
 # Get environment variables, readily decrypted by rultor
 source ../rultor_secrets.sh
```
Comment (jayvdb) [inline .misc/deploy.pypi.sh]: Why is this line running in non-rultor mode.
It obviously was not tested. Please label your PR 'wip' until it is working (or test on your own travis on your own fork on a different branch name).

**Comment by Shade5** [comment]
Now I should fix this #3644 right?

**Comment by Shade5** [comment]
How do I stop git-bot reminders?

File: .misc/deploy.pypi.sh [hunk 11]
```diff
@@ -3,25 +3,44 @@
...
-if [ "${PIPESTATUS[0]}" -ne 0 ]; then
-    SEARCH_STR="500 Server Error"
-    if grep -q "$SEARCH_STR" twine_output.txt; then
-        echo "Server error 500"
-        exit 1
+    # Ship it!
+    echo "Uploading coala to pypi"
+    pip3 install twine wheel
+    python3 setup.py sdist bdist_wheel
+    # Upload one by one to avoid timeout
```
Comment (Mixih) [inline .misc/deploy.pypi.sh]: hmm, I don't think this comment is applicable anymore as they are all being uploaded together

File: .misc/deploy.pypi.sh [hunk 12]
```diff
@@ -3,25 +3,44 @@
...
+    twine upload dist/* -u "$PYPIUSER" -p "$PYPIPW" 2>&1 | tee twine_output.txt
```
Comment (Mixih) [inline .misc/deploy.pypi.sh]: add `-r test` to make sure it goes to test?

File: .pypirc [hunk 13]
```diff
@@ -0,0 +1,6 @@
...
+    test
```
Comment (Mixih) [inline .pypirc]: can you add a `pypi` section too so we can upload to the regular the place the regular way?

Comment (Shade5) [inline .pypirc]: There is no need for a `pypi` section. It's `pypi` by default.

File: .misc/deploy.pypi.sh (hunk 11 above)
Comment (Shade5) [inline .misc/deploy.pypi.sh]: Right! Will remove it 😄

File: .misc/deploy.pypi.sh (hunk 12 above)
Comment (Shade5) [inline .misc/deploy.pypi.sh]: No, rultor has to upload it to `pypi`, which is the default.
Travis has to upload to `testpypi`, which has been added in the other if case.

**Comment by sims1253** [comment]
@jayvdb you didn't give a reasoning for it so I removed the blocked label.

**Comment by Shade5** [comment]
Actually, this was blocked by https://github.com/coala/coala/pull/3670. That's been fixed now, thus this can be merged.

File: .misc/deploy.pypi.sh [hunk 14]
```diff
@@ -3,25 +3,43 @@
...
-if [ "${PIPESTATUS[0]}" -ne 0 ]; then
-    SEARCH_STR="500 Server Error"
-    if grep -q "$SEARCH_STR" twine_output.txt; then
-        echo "Server error 500"
-        exit 1
+    # Ship it!
+    echo "Uploading coala to pypi"
+    pip3 install twine wheel
+    python3 setup.py sdist bdist_wheel
+    twine upload dist/* -u "$PYPIUSER" -p "$PYPIPW" 2>&1 | tee twine_output.txt
```
Comment (Mixih) [inline .misc/deploy.pypi.sh]: hmm, we don't want automatic upload from rultor yet. We are not ready for that

**Comment by Shade5** [comment]
@Mixih Did you review this? I can't find it here. My mail said that you said this `hmm, we don't want automatic upload from rultor yet. We are not ready for that`.
Actually [rultor](https://github.com/coala/coala/blob/master/.rultor.yml) already automatically uploads to pypi.

Can we get this merged before the gsoc deadline :smile: ?

**Comment by Mixih** [comment]
@Shade5 yeah sorry I messed up a review:P I'll go ping jay since it looks about ready

**Comment by Mixih** [comment]
@jayvdb ready for re-eval.

**Comment by Shade5** [comment]
@Mixih Cool, thanks a ton :smile:

File: .misc/deploy.pypi.sh [hunk 15]
```diff
@@ -3,25 +3,43 @@
...
-# Upload one by one to avoid timeout
-twine upload dist/* -u "$PYPIUSER" -p "$PYPIPW" 2>&1 | tee twine_output.txt
-if [ "${PIPESTATUS[0]}" -ne 0 ]; then
-    SEARCH_STR="500 Server Error"
-    if grep -q "$SEARCH_STR" twine_output.txt; then
-        echo "Server error 500"
-        exit 1
+    # Ship it!
+    echo "Uploading coala to pypi"
+    pip3 install twine wheel
```
Comment (jayvdb) [inline .misc/deploy.pypi.sh]: these rultor specific bits can be moved to `.rultor.yml`

Comment (Shade5) [inline .misc/deploy.pypi.sh]: Yes, but shouldn't uploading to pypi be in `deploy.pypi` ?

Comment (jayvdb) [inline .misc/deploy.pypi.sh]: Yes. But this script shouldnt have 'if rultor' or 'if travis'.
The script should work on my computer, if I give it appropriate args or env vars.

File: .misc/deploy.pypi.sh [hunk 16]
```diff
@@ -3,25 +3,43 @@
...
 fi
-rm twine_output.txt
 
-echo "Installing coala from pypi"
-pip3 install --pre coala==`cat coalib/VERSION` --upgrade
-echo coala versions: pip=`coala -v` repo=`cat coalib/VERSION`
-[ `coala -v` = `cat coalib/VERSION` ]
+if [ "$CI" = "Travis" ]
+then
+    # Ship it!
```
Comment (jayvdb) [inline .misc/deploy.pypi.sh]: this is a copy and paste of above. 🚨

Comment (Shade5) [inline .misc/deploy.pypi.sh]: @jayvdb So do you suggest removing the `deploy.pypi.sh` file completely and replacing them with two separate files? One which is called by Travis and another which is called by Rutor?
This way you can mimic their actions by calling them on your local machine too.

Comment (jayvdb) [inline .misc/deploy.pypi.sh]: No!
`deploy.pypi.sh` should be a generic script that does one thing.  It should be able to run it on my machine, or use it in Rultor and Travis.
Any difference in what they do is configuration (e.g. via command line args), not `if` branches inside the script.

File: .travis.yml [hunk 17]
```diff
@@ -81,6 +84,9 @@ script:
 notifications:
   email: false
 
+after_success:
+  - bash .misc/deploy.pypi.sh
```
Comment (jayvdb) [inline .travis.yml]: for this to work, some travis secrets need to be added to the .travis.yml, and also we only want this to occur on master, not on PRs.

**Comment by jayvdb** [comment]
Please feel free to re-claim the issue if you want to start work on this again.

**Pull request closed without merging**